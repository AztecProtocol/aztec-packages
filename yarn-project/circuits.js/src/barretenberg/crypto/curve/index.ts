import { Fr, Point, PrivateKey } from '../../../index.js';

/**
 * An interface representing curve operations.
 */
export interface Curve {
  /**
   * Multiplies a point by a private key (adds the point `privateKey` amount of time).
   * @param point - Point to multiply.
   * @param privateKey - Private key to multiply by.
   * @returns Result of the multiplication.
   */
  mul(point: Point, privateKey: PrivateKey): Point;

  /**
   * Gets a random field element.
   * @returns Random field element.
   */
  getRandomFr(): Fr;

  /**
   * Converts a 512 bits long buffer to a field.
   * @param uint512Buf - The buffer to convert.
   * @returns Buffer representation of the field element.
   */
  reduce512BufferToFr(uint512Buf: Buffer): Fr;

  /**
   * Point generator
   * @returns The generator for the curve.
   */
  generator(): Point;
}
