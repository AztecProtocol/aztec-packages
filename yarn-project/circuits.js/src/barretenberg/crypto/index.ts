export * from './aes128/index.js';
export * from './grumpkin/index.js';
export * from './pedersen/index.js';
export * from './ecdsa/index.js';
export * from './secp256k1/index.js';
export * from './schnorr/index.js';
export * from './signature/index.js';
