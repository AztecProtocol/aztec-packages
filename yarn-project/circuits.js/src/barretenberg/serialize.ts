// TODO find a new home for this as we move to external bb.js
// See https://github.com/AztecProtocol/aztec-packages/issues/782
import { Buffer } from 'buffer';

/**
 * For serializing an array of fixed length buffers.
 * TODO move to foundation pkg.
 * @param arr - Array of bufffers.
 * @returns The serialized buffers.
 */
export function serializeBufferArrayToVector(arr: Buffer[]) {
  const lengthBuf = Buffer.alloc(4);
  lengthBuf.writeUInt32BE(arr.length, 0);
  return Buffer.concat([lengthBuf, ...arr]);
}

/**
 * Helper function for deserializeArrayFromVector.
 */
type DeserializeFn<T> = (
  buf: Buffer,
  offset: number,
) => {
  /**
   * The deserialized type.
   */
  elem: T;
  /**
   * How many bytes to advance by.
   */
  adv: number;
};

/**
 * For deserializing numbers to 32-bit little-endian form.
 * TODO move to foundation pkg.
 * @param n - The number.
 * @returns The endian-corrected number.
 */
export function deserializeArrayFromVector<T>(deserialize: DeserializeFn<T>, vector: Buffer, offset = 0) {
  let pos = offset;
  const size = vector.readUInt32BE(pos);
  pos += 4;
  const arr = new Array<T>(size);
  for (let i = 0; i < size; ++i) {
    const { elem, adv } = deserialize(vector, pos);
    pos += adv;
    arr[i] = elem;
  }
  return { elem: arr, adv: pos - offset };
}

/**
 * For serializing numbers to 32 bit little-endian form.
 * TODO move to foundation pkg.
 * @param n - The number.
 * @returns The endian-corrected number.
 */
export function numToUInt32LE(n: number, bufferSize = 4) {
  const buf = Buffer.alloc(bufferSize);
  buf.writeUInt32LE(n, bufferSize - 4);
  return buf;
}

/**
 * Deserialize the 256-bit number at address `offset`.
 * @param buf - The buffer.
 * @param offset - The address.
 * @returns The derserialized 256-bit field.
 */
export function deserializeField(buf: Buffer, offset = 0) {
  const adv = 32;
  return { elem: buf.slice(offset, offset + adv), adv };
}
