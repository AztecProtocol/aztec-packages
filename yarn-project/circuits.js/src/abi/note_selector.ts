import { toBigIntBE } from '@aztec/foundation/bigint-buffer';
import { randomInt } from '@aztec/foundation/crypto';
import { Fr } from '@aztec/foundation/fields';
import { hexSchemaFor } from '@aztec/foundation/schemas';
import { BufferReader, TypeRegistry } from '@aztec/foundation/serialize';

import { Selector } from './selector.js';

/* eslint-disable @typescript-eslint/no-unsafe-declaration-merging */

/** Note selector branding */
export interface NoteSelector {
  /** Brand. */
  _branding: 'NoteSelector';
}

/**
 * A note selector is a 7 bit long value that identifies a note type within a contract.
 * TODO(#10952): Encoding of note type id can be reduced to 7 bits.
 */
export class NoteSelector extends Selector {
  /**
   * Deserializes from a buffer or reader, corresponding to a write in cpp.
   * @param buffer - Buffer  or BufferReader to read from.
   * @returns The Selector.
   */
  static fromBuffer(buffer: Buffer | BufferReader) {
    const reader = BufferReader.asReader(buffer);
    const value = Number(toBigIntBE(reader.readBytes(Selector.SIZE)));
    if (value >= 1 << 7) {
      throw new Error(`Invalid note selector: ${value}`);
    }
    return new NoteSelector(value);
  }

  static fromString(buf: string) {
    const withoutPrefix = buf.replace(/^0x/i, '').slice(-8);
    const buffer = Buffer.from(withoutPrefix, 'hex');
    return NoteSelector.fromBuffer(buffer);
  }

  /**
   * Converts a field to selector.
   * @param fr - The field to convert.
   * @returns The selector.
   */
  static fromField(fr: Fr) {
    return new NoteSelector(Number(fr.toBigInt()));
  }

  /**
   * Creates an empty selector.
   * @returns An empty selector.
   */
  static empty() {
    return new NoteSelector(0);
  }

  /**
   * Creates a random selector.
   * @returns A random selector.
   */
  static random() {
    const value = randomInt(1 << 7);
    return NoteSelector.fromField(new Fr(value));
  }

  toJSON() {
    return this.toString();
  }

  static get schema() {
    return hexSchemaFor(NoteSelector);
  }
}

// For deserializing JSON.
TypeRegistry.register('NoteSelector', NoteSelector);
