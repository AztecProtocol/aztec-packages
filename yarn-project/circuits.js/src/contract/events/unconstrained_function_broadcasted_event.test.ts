import { FunctionSelector } from '@aztec/foundation/abi';
import { randomBytes } from '@aztec/foundation/crypto';
import { Fr } from '@aztec/foundation/fields';
import { type Tuple } from '@aztec/foundation/serialize';
import { setupCustomSnapshotSerializers } from '@aztec/foundation/testing';

import { getSampleUnconstrainedFunctionBroadcastedEventPayload } from '../../tests/fixtures.js';
import {
  BroadcastedUnconstrainedFunction,
  UnconstrainedFunctionBroadcastedEvent,
} from './unconstrained_function_broadcasted_event.js';

describe('UnconstrainedFunctionBroadcastedEvent', () => {
  beforeAll(() => setupCustomSnapshotSerializers(expect));
  it('parses an event as emitted by the ContractClassRegisterer', () => {
    const data = getSampleUnconstrainedFunctionBroadcastedEventPayload();
    const event = UnconstrainedFunctionBroadcastedEvent.fromLogData(data);
    expect(event).toMatchSnapshot();
  });

  it('filters out zero-elements at the end of the artifcat tree sibling path', () => {
    const siblingPath: Tuple<Fr, 5> = [Fr.ZERO, new Fr(1), Fr.ZERO, new Fr(2), Fr.ZERO];
    const event = new UnconstrainedFunctionBroadcastedEvent(
      Fr.random(),
      Fr.random(),
      Fr.random(),
      siblingPath,
      0,
      new BroadcastedUnconstrainedFunction(FunctionSelector.random(), Fr.random(), randomBytes(32)),
    );
    const filtered = event.toFunctionWithMembershipProof().artifactTreeSiblingPath;
    expect(filtered).toEqual([Fr.ZERO, new Fr(1), Fr.ZERO, new Fr(2)]);
  });
});
