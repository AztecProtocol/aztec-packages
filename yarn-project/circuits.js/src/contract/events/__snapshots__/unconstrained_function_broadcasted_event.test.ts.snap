// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnconstrainedFunctionBroadcastedEvent parses an event as emitted by the ContractClassRegisterer 1`] = `
UnconstrainedFunctionBroadcastedEvent {
  "artifactFunctionTreeLeafIndex": 1,
  "artifactFunctionTreeSiblingPath": [
    Fr<0x28914d31533b25309c8e37f0eb04595c4ebcdb749c74ed4d344e19a00bccb0e6>,
    Fr<0x0000000000000000000000000000000000000000000000000000000000000000>,
    Fr<0x0000000000000000000000000000000000000000000000000000000000000000>,
    Fr<0x0000000000000000000000000000000000000000000000000000000000000000>,
    Fr<0x0000000000000000000000000000000000000000000000000000000000000000>,
  ],
  "artifactMetadataHash": Fr<0x2ee0a6721704053281d575f82984fad809a7cf5b1f2d47971e1429fbe2eb1578>,
  "contractClassId": Fr<0x06e2af08f53b9a05db4870972b4a9a663a5ffb54e962020d4efe3f1b9a21343f>,
  "privateFunctionsArtifactTreeRoot": Fr<0x2b8dfed76d4d8bb88895927b771bba06907cefe8908604a485d34d39d88c66f5>,
  "unconstrainedFunction": BroadcastedUnconstrainedFunction {
    "bytecode": Buffer<0x1f8b08000000000000ffed9d5b8c24d759c7ab677aba67a6a77b76767766e7b2332e2305450a123dd3b3761ce5a113040a020564a3480829d12ebb0e8e8c97d863122362c4ed21912081806c88201188f0406cbfc00b029470110e128ea548e422140b2c0c444402e13c6615576f7ddbbffee6abeaae9ef3d554db53d268aaab4ed5f99fafcee5ff5dce39b5e8f6b1fcda5f2d3d5f50ff79489a7efabf7bb2633fe0bbba9e386b3382736e4670ce07c499609b8f7ceb69dd41aea1312ecc00c646e0ef2e18ebe979f3b5bfc5d7fe96a2d13eedce11aa203a530a2c6984f3e9fde4a334703f0be85bf0bc1cff8ef3adf4ff3b1e7df4ea13f1438f5cbff1e1f8e6e347f1cd07e36b371f7fe4fa637cf0bfa77df07fa77df0c3f3533ef8db78f0eef4ffd5a3a31b3ffb7347f1d1cdf8b1c7af1d3d7af5a78fe20f3d74f433f1cd9fbff1e8830fdffc105ff0f4b439ffff4973feceb439d7eb533ef8fd78f0f271c857af5fcf46fbce69337dd7b40f3e7a02b4bf3a6da61f9df6c1cf9d00eddf4f9be997a67df05fa67d90acae6831b717a6ccf4074f90e9d569337d6ada079f9ff6c1574e50cc071b5366fa678de933fdab6933fddb691ffc8f13a0fdceb4992e36a77c7075da07ef6b4e5fcc07a6cdf4c91364fad16933fdf8b40f3e7302b49f9f36d37f9cf6c1178a3c7887c47d357d6a116fea478148e85bbbd71a7839496872b4a2db449400926b425a6ab82683c91cae495f36bf387c45237c190e17813dd87b5f23e8ad61b1efe06ee15cee25796fe35a3d7c19ef5d843c83bdb7eb26bb816625f211adb289df22aba5f43ca075649f79d7d2bf25e419a5df48ceb79056d2893ce6f08d93a31d0dbfff52ce735ab1eb204dc3287f3f0a5bfea6c2d3549817524cc9b196dc5f1cca45b0b5d2f35b5362bbf6d6836bfb37f6efbb72bd7bf5e0c12bf75def1d5e49deb10c0c521724af3aeeff39d2ed0197c82fa4458ded2b745d6ca26c35e0673b98c3b70a93f7fe21fbe50865e3d1c7f91cf0cc07c7d3ddf729e7ed6fb70cec61de7be54a22ab9692555dc9aa8334c4d072905f0df9cabbe577cbc83b9c2ceeb991e4bd32812c560c3c2b25cb62c5c83ba02cae2579b7279045dbc0d32e5916925f51ccad0a605e54e761f2bee7c15a341cfbf264d131f0744a9645c7c83b9c2ceebd37c97b750259ac1a78564b9685e4571473ab029817d57998bc0f063cfbdc04b23867e03957b22c24bfa2985b3388b95d01cc8bea3c4cdef71e2679af4d208b3503cf5ac9b25833f20e38960c38e2f9096471dec073be6459487e4531afce20e67315c0bca8cec3e47def7d49de172690c50503cf85926571c1c83ba02c06e3dfc5096471d1c073b16459487e4531afcd20e6d50a605e54e761f23e1cb4bff50964b16ee0592f5916eb46de016531e0021b13c862c3c0b351b22c24bf049fd8e43ee58a67ff4a4be1498e3c1b9c6068e17c0e182f05c7d8edf2fb5c425e9be1f3dab7ea8bfcde74cd7b7fe0ebd80e5fa6c118b493be4b7c1ddb4699c44719dabebc83f7d6908f5cafe3fc69a49574220f69b7823dd161c4377239e7b90df55c0769b68cf2f7a3b0e5df5678b615e6e49bfc067078d4ad49ea75c249ad7ed9a11f1cd8feb53f77927ec7a98fd9772a67973e8569fd54d651c37713cc4d25ab0ed2b08f3ead7ef30cf319e62ccc8c6568e29af883a5cd277de5d7e787781dfc5807c93b97a3d123af6f72f645f45ad1a89f6f1c1edaf63c7c5b4ee5ecd28710baafd4f6f8969255d93e04edeb92df793e8433cce1312718a4adb7f05fd22d57046399bedaacd89cb66fde5dc6e4c9fb93df5f46be0efd598f7114a21f2d2879d791e6979bc3b45f015e1d6791558fbc7cb459f5887ee2159c0b9e65273c526f449f6c1ab2081e6fd91d8db709f8de1e652cf5a4a1ca54479a97d3fff4cb459089fe5e527fe49dedc86e7b1edfcba95ef6382688cc047f07329334df1c23b3acf1651972d2fda753d90e9c7ca73dfa2f45661d25973ad2fc9fbaa765b6962133f60b222bdaf23d62c49277ae18182385510e726b0fdf6c516e4df978f8cd5bd1e8f71a87c7db8fd88a8ed7ab3c3cdebebc5634ea671d87c7db9fd68a467d9de3f078fb5a5ad1a81f6c1c1ec1c0d848c6845bfc5dea27f9a9d491395c93ef348f6b222bc6715f00ce7e14440edd2c9faa477f5d54e6de6dd6a99c5d6b2cb9a8cac4fad14439bde26a74cc93fc5ec3772882b93d8398572b8099752e0e9677ef1e271ed24b64b60cb9b6947c592e07bfc041511e44ff87c7b8e1e5ff48ca744995694595a983341c173cfc3c79b6e84bf80e4530af550033bf5f1c2cefde15279fc081b43f91eba6922fcb253eda9036617eb744c7128e427f30fdd681f21df1078b3f5cf2113c75a4797b6d88f16de939ed64f48bbcbb36c4eec0750f18f311a10c3cfa38f7e6ba4ee5ec5af18b1baa4ce4b2f328a757fca28e6dd3dcb628e64b15c0ccef1787cbbbe7c4abeff45922d7f34abe522eb64fb9a6dba783cc07ed93717fc9719abab15339bb56acf5ba2a13f5cf3994d32bd65ab73d1ddb5c14f3420530f3fbc5e1f23e74d253eeb44f91eb9a922fcbe5e1674ede710edf8f768fe4a8e3fed318d7aff9f60bbd3752bfa0f566ab8d95699fc86a6379ba7e1ee632ec13a13197619f1887d9a72fbb6d9ff0b029eabeec9c922fcbe5e1db6a45a3beb6e4c8eb3318cbe1b1d68d5339077d865e3740c7a7d0364efb84d7ba011d85477eafe03b14c1bc5601ccfc7e71b0bc6fdb273c6200a5fde939da56b91cfcf10745d746614c85471feb54ce41fb6ba832d555993a48435ddb217664509797151ef9cd757d8a605ea900667ebf385cde3d6ffba0c87553c997eb9d6da86b09a7fe9b334e1de2bd679cbaa298cf3875f1179f71ea91e38c5347679cfaf871fa9c9a63ba5cd363ba87fdb0681c8fd50f06c4b3ef544e734d26bd968315d3e5d42fef5b639f5e9ba728e68d0a60e6f78bc3e57de83417e540c7c4e89860ae77a5c7d0a47dbe68f89f43dad7d91f96e1bfe4d1c739f9981ced8ae1d9a8189e858ae159af189ec58ae159aa189e32d6882a82a756013cade8383f613c3ad74a953e9cf18e73c6fbc4be467dfe2f9ac3fb3226cce11991cb3cae696e4f0cab06d60eaec998b262bc8f72f79a47c7bcfaf8cdf9323585f1b4f19caf189ea58ae159ac189ef58ae159a8189e8d8ae169570ccfa572f01c700e73045c3cfa38779ed37fc0f166123cb4937acd672c62d3e37c468f79bc45e7333acfc5ee155d4b85fe2a2f9ba7d79c73f151855ebb644bc94afbc03a48c3f5dcb61ce467f93ee537f7a939c3ec8f99eb8270cd09ae1f50058c72ade98be7b0e858c5b91e1e7e79a77276398744dedd36e46eadf1e8b0bee4bed576e437e7ba9c61f6c7ccbc4f300ef5f405ce53ca93c58e8167a76459ec000f397372b414de2a60946be43ee1fd7ab7ede7796b6fd502e7253c941cdca13f3c28ca799dd77cedb18e4d82e732f078d443a7720efa845d5526bd1e6f0769d83fee3a94b3161d5ffb577eefe23b9c61f6c7ccf62f585b48d7aa0846bd16b5139e8396c2931c797dc22ef0dc151ecf802bef16c07317f0ec85c7b3ef54ce411f15ab32edaa327590866d277628670df9cabbe5778cef109f61ae1ce6ed0a60669f25585b4877b92218e5da9e339e96c2931c79fd586c602417bedb09635c00e3ddc025cfadfa62ec32e64ede9ff8849e9a1be6eb302e0ed63adc437e7de090fcea48f3ae8561da4fa5d8dab8cf357c77d4354fbd5cebc65a0f6ea33ce41a5e7aa7e68b3a6ff2b3bd8aca8cf6845dc851ae914b6a3e4c7b7c3f30ee2c9d7e1318a50cd435bd6cb059fb8bd0cea1796ed564463b88b5bf4b8c6bf20c39e93ad2f5c3e01e7047eec914236fc9af168deea3c0b2f25949f397699f95f45f5f9c1b96dd6a8bd49ba49c55f83e79fad32cd42bb6c5381c9e2e39c59651f6d0f6b558959575ca6a3b09968df058f66baa8c6c3782816de005b4819bf3c7e5d48fca190b5947acf6c7bd49b4fcfa918ffce4dd7a3db40eaed136e0653bc91a53c6b57db9cfb6ef612fc8fbae7bc028d72cff481c0e4f97fd9ebc3fa9df2f19fc3960bec7f8731c1defa7c99fe7c19f5ff6c53690c99ec292e4fb0af2dd7192c9ae9289e0e09e7392e6b9fa30ed37d17f52bf96ffdbeada69f23a7261722639bf0bd7f2781479f4babac6b91cfd286cf9f41ea57a3f52c6015afb84c6e1f078ad4975f806995376b89ae2dd54e590b2d78defe8359666cd5be6582aed976d9c7b9b0ac6d3da0fad836bd4d92f2bb95267679bf5e205593a3b79815c636cbff445d4ede3f49c5cd9c1d67050d41f4baee0d0f7f5388e4c8287eba87af5091eeba8727c0f1d13a9b9dfb69215b99fb36f2f77add33c1fee19e6f0982d7d9f63f056453096b416ee080797f7273cf7cdd0bbabc0c11f0707ffbe141b39386d61966dd56b9ccee2e0b4f55a1c5cdbeeaa664f684776dca2b71c6b39729c757d24341e6dab2b33ef2c9eb70e59c839fda856cc8fde63c38bc7e4f5bde4a272cdda0f9e6d54efd37eda3e1cb9463d46646fd5895a408ce4eb960e17473e6396bc3f1917ee9f3ffe1d02e6dbe377d06316edd892e60063d67b3066cda22f7a56fdaa966d6b56edf96736f591a337a9dd807ea459b5db7acd15726aaf3df21389dbd1edb58e344fa4df276b8f526d0fdb3564a6c73f27dd659fe3cf0ecaf864c574a697e687697f658af1c7abfe678d3ffc96938e3f5a27386d9dc91a93e43cc6357986312a0efafd88de22f1ac7adf1fdac41b699a8baa6c75a4f9644e5b9d8b8eeb21fd6894d387b4f5717e523f0adbd624ee53c71dae422692e6f77264c2f95afd80189dca3eb2577c3d1aed6f597649f387e8673e939e733d1ffa9f9f33eecb9167e3660caed73ae34de0ec47c7f721e73ce0d0f1314dbcb7161d9f135dc7f9b3f3c3b47a7eb5c89afb9dcb1a2079eb93afa9e7382fbb0aeb9327dfe48f51cf9ec358ebb57ee34a868cc8d1258de5cbe2dec9717a5ec6bc06ae89d98f46f75d153c32ce4b9918835e867f53f292776bf950665c2fd3eb5b6f29996da9bcb3742fb94ffe715a7bec727d53e739ac876fa439ac5aeed6dc4a49c37547bdea6ad6bcc5bcb570f3302fcc20e6f3338879bb02985bd170ad3bdae0653d3eeab35f0757e3da2c724dc6e6255c9b4bcf9bb826eb3e323d753c192f6b2aff383a1e1f58479a97c107de92da5ab3ec3b558a95a72e6bd929a9df6ea86b65f893e4dd5a8fa32e4b9d57ebb2f4fdc4c0edb057e4884f477885d419ee7d2b63a4e8b7724fd2d691e6d51c5dae1d8d7226f9afd7ac77d2f906e5e57a72fde8f89a1d6d60e77aacabaaccfd14673b3cce1ef9f382c249dd52d2dc1a63ffcbe2e27ca7de278065bea8caecc0d172cbbc669479a19e5de6c5c865df889e273fd536036bdff23beb91c137b6929ed3174c3be78e715f8e3cce49bdd561adc091bdc9e47b378dbcebc01a28ef91b5d5c46620f9c8f53aceb7ebc3b475252f91b5604fbed132de91f5dc45f55c273abecf19cbdf8f7c6c26f26eda4a9223f926aba867528f3cf9d772868cd8476bbb12f571c64254694ce1da50964e4c7e65e9c45e3139593a31e30fe51a6d0bc25568cb89d373e12ee4955ebe24ed534ddead7d3ae4948d289b174b9a6eceb89235e7e649f887b42fb48ab169797e96598a0fb3e2a4dad1a82f2839122ea3396b3f72ebcb7aec13646c93fc25bf3ad2fc404e9db3fa17a97bdc8bc0dabbcec36f42aea2cbd641d924cd0f8fe1691efcd9b3afe7de0bc97bb78cb24b9a7763fcfc71f030f9468c877a9f715f8e3c9ec67568cbe669ccbb0a3cedbd393c4d645d94a7adaae7aac8d31e403d7bdf29f0349111fb014943fbba5eb3927ecc32e257b3f614b3f600b5781ac7718ba779f993b3781a79a6de9b6d9c2f8df39e1ce67c1e8b174edebda7f0d06ea6b919e7364a9ac726e4669c8b71ffeb60be838e959e05dbe61cce35e772e090878cc3d13127ac4b92e6d7c6702e3dcf55c7c7917b729e91c77ed86c4bba6c17503649f3b1319ccb63de6519fdb6b68db1ec92e6e3180b7f0b9c4ac7030fe26d8cfb72e4712ef69d0e7d7ed78a9f5a35f26e036ba0bc47f6c615ce25f990e3caf9a7c1b9249dc843c73dd1b64eecfab94df51cf762386794bf1f9563cfa1def33ba8679f01e7f29a77702e43461cef250dc77badeb73ece098e7b5565ed6d8c179f7728d3e5ec67d86c4c3313879f79e924f1e279167c9499e99909350d77bf38cc7d3677192598a27a6bd88f3ace4daba51be0b46f9bcfc4159eb6cac01a39e7f57b29debb0a89deb1f0adab934bfb2ec5c94c3389fb2973e9c355e70ac5b57b83d6d8f9a3b58df44d2bc38e69b7454d9749d6b47c7fd1d51346acfb56c935e3a4056d92f1a65ffca189eec60433c2cc39fa36d932cbba4f957f0976f8cb14d7e6b4adb2463bf1dfcf15ddaa2e47baf1879333e3c50de237630e1c9920fed8a72fe3fe0c9daa627b216ecdc1393d8f573abeab94e64db663df6fab3e24a5614e6e49bfc1bead9b7c093bde328b48c68e39334e4c9d69a55eb3817dc5e3c2ccbff45ff825cb3d6ace27857d535ab680b16fe65c5ac730ea2971e40bd2479b7a507e8b8373d6f9e76875a1a083c4e0fe0bc4e59ab91f30a19cfa9e72b4a5de0bafff473f7d3f3eec98e817cf6f05e7e676bddcb25ac7bf98e85e3657a3dccd5af9a6e93b7c64b9e6e935439ddb7f4a3514e167a8e61593e762997a57bece4b44faf79909e3a076d71c97b778db2dfd9e70eedf37bd273eab39c63b86fdc97238feb91db97cdf5987715b85e776198567336917551ae77413d5745aef726d433a94765723d9111799da4614c9b8c3d923e8beb7971a62cae47ce24d7c899ac3974729fe35215fce75c536412ae47ee40d9c781cbb0170d8fe4dd7a3d6ff2864664f32ecea3f9a131bc4f3f9bb56e793f0a3b3616d9f7e747d16effb431c4e6c0cb7a5c73478ebc71c57baf5ba7729a7bcbebb57de89fe1ba5fa765eb2f8af95c053033ef70bcb57768f1704b169b069ed38a1ff16ab39485ee4fe8139334d772fac43cb952bfd1fa38f51bea93555aff907a19fdf45ef6deacfa60719dcd8acacc5a33923e1deae016d7f18ab9cbe23af43b699fdc69fa03b3d6afaf92ccc80f2d9f5e0c8c8c41e907c4c8fdd8a947af1978b8be541b1803e1e9693d3e563804a75c9b8b46fda47c56d27c1b5ceed7618b239f9632e9753b42c97847c958736c6bfe3adb4e1c58c65b4ac69237dbedaa932c76952c241f8b73c9f7b56cad92e60bf8be9fc8f9be5ccbee9378e6f3868ed58fc2dafbbce6076b7b9ffe8eb4f73d7d4af63eaffe5fdbfbb68db24b9a3fc0f7fe34ec7952afc8e59e35eecb7166ef9bccdef74c8ebd4f645dd4deb7a39eaba2bdef8f50cf9e45bfe2a50bb6336444bb99a4e19abac2736937b362ec4e4b6fe3fae3d6beb2e473d69ad15eb19b596bb1d0ff2cd7ac18f838201eda11ad7d3d02b6ffae354f866367f2ef5563eca5df8df39de5da964a978cf7d27f6b5f5c597337a93f2407e76e7e71cc183aeb733769c791eb92e6057ceb173146ea98d2813dd7b82f47de18eadc8fcfd4dccd6f600cd5733045d645e76e6ea9e7aa3877f3cba8672f610cf5d2d1973364c47e40d230cec4610dfa633683e4dddac64a7b6123b2f537da1eff33a7cfcada4345c7e07b728149f64bd6f366b2b880dc2f630f95ac1867f2152bd64ec63ad6af383de77a75afa21dfc267c505edc268b9391db58b6d32aadf5ced83a6bbf88d078b26c90d65e3faca7d65e3f55e0dace9cb5cbbe8573531b868f3560be3dd61d891dd4fd2a7dd86f6f0ed32e1b6d2f20b611991063bb6232f9afc630ed5a7a4e5e5fe5f531f3f6f1a18e62edad30eb7e97381c9e43271de7f0246bbc78e8465ee5947dab2faa72701f04fd1dbdfc90223779b7fc2657b1ec448c11138c5efedcacf195fedcbc31977300ac31f734b994d6b5c90dc967e2f49c310d0efceaa0e89c7fc67778f80e8ada0e9cd7f9daf75adb80e37be87dab355fde52b2cae2cb5eed224b67a27ff40cb33f662b4e8b63f06645306abf3edb601c0ecf0807a74ff5c72ac6c19f05077f001c7c56f720f49ee724f6db93acd7e0c56db2e634911b6a5da60c3d4a6436c9de734ef229751f36bda632e306de9fb6b1d7eb3e6c3b46d925cdc3e86b1e49cfc9ff3917f323c67d39f2381bfbf537fa3e6cbfd818a6d5fba989ac8beec376413dd789aab70fdb0751cf3e82f1d62b6e68254346d429258d153341fd57cfe7f5e44b7b0ab75ed7bb836b560c2cf94055e748597318e2807826d5b1e97399d579db9e7591319c7d236fb697bd8aca8c3a9915877bd9c07db902b8591fcbf8d65cf7a01fd9fdce6525c7aac98cdf9af30039b7cf6b6eabeeb7f5bc43e26fa834d6fccacfe570d296f16c52cea7e66e9f530fa30fd7d233bcf49e2c7f0df59e3c7fcdacae3fc7fae5c4f173d7fbb2f49bbfcea94b968f40c736726e08f9b457ac5ed6da7a8c5793347f374677f38a77f78ad5d3ba9b15ab27699e07a7fe27e866523f6973fa9a715f8e49eded65eb6eccbb0abadb57737437917551ddeda27aae8abadb3fa39e7d0d63a9979d71254346ec07240dfb256d5ba30ec431af6a3a10f7ed0989876370f2ee5d251f62d631869c3b26695e19c349f4b3b4adbf1e6248c84974bcc869cfbfb6ecca72ced817c67bc9b535a37c1da37c5eeb8e67c528b4f15ffb88e670ae3997c3bcac43f6950b0a27e3eb25cdad319c4bf7bb9a5f71bed4b22107b6a16575cd4906b9e305e779ad29dc4e787a1677b0be89a4594a4fb2be49d63aa3fc265226a665bfdf36caeea50364957dd528fbb99cb23bcdc338f4f471e8392d9b46d925cd3ae23c2fa5e76c371cafbed7b82f471e4fe65ee70ef32a063c791e38fbc88779cf016ba0bcf799b7f064c947aed771fea6e630ada4137990572447d25e846b12bb7e4ecf7be9204dc3287f3f0a5b7e3dc7a6ae3027df641bf54cea91e7b8d9c8901163cd248db51f19f95807e782db7b4eb7bcbb6360946b56bc1fc7bbaac6fb71ef18e15ff43dc5e9b935a7330e8c9b36ade4dd961e2078b51ec0bd8624cddb72c692acd8ffb6a10770cd5faebb1e47b69fa3ec753bacd89c77a28dff42f37899ce749b62b83d759ba46f963e25cb2e1b7acd5f2f9eab750f2997a57bfcc418aee760c7eb79ea1cb4c525efdd36ca2e697e12edf3a7c0e5643ca14df403c67d39f2b89ef31cda5caec7bcabc0f51ecae17a3b489f1c9372bd8e7aae8a5cefbda8671f3805ae6771264943bdb585f3281ab58996e19fcfb28992d7e9b5bfc8f53886ca7d8e4bde7e187937b1c97fb9c6399693703d7207cf39e6f49727efb6e6988b7c35ef63db9534bf3421ef631fdb00ef93fa46fe6ed94b4eeb9bd267e33cf7ccf4732e283cf4737e6c8ced4ac7d9d37f23e5d1f3d01623bf3d0ad632cac675fa25cd27c6f0158f75d59cca3ea867e751c6e4bd2b46d925cdef621c790a7c44db7993fb9f35eecb91c75738cfd0238e3bc97b1d38fbc887796f006ba0bc47d69517bea2f769aee3fc4fc057f41e9d226bea19d26713bb7e4eef51db898eef53ccf2f7a3727428ae31fffba8679fc5d8e3d5d75ec89011fb38ea7882a7e58447fb940587e4a76d204c433d43d23c9bd367712d19b6636bac2b6b8f08f94dbea8c7bf9651e6c15a3dc0eb10873098034ace9d1c79fd1975800507f925dfe752faaef7df38fa911b4fbce7eac30f5dbf7af4d0cd47eebff1c1c76f3c7654034c813eafa0d7a2d162dc421aaa3c3cb7eedfc2ff45a4e94761abcc02f2ece3b7e4e7a47275bdaa95e522d36a244d075f28e81ed46a611bb2a24bc6237467317db787cc9694ccd8dc446692e6f931325bca90590332bbe39e453e0e653bf0327d25e56c2999097ebaad24cd97c6c8ac9521b345c84c644573b36eaf2ddc6fe29a60e23b68da906bf3461eda0c95bce3bbedf960600d5b0100>,
    "metadataHash": Fr<0x1a8b89c75f0ad79c19ba53a8bdfcb25bd4f8d33bfa11f1f52528a8ee58fc2543>,
    "selector": Selector<0xbeda04d2>,
  },
}
`;
